### PR TITLE
Fix #14046 middle button zoom level

### LINKED ISSUE #14046: Darkroom: Middle mouse button zoom level
First of all... Awesome program. Thanks devs.

**Is your feature request related to a problem? Please describe.**
I'm always frustrated with the way the middle mouse button cycle through zoom levels.
First 100% then 200% and back to FIT. When zoomed out (< FIT screen)  it cycle through the 100% 200% thing again. 

I have changed the source code in my personal build to fit my needs but I would love to see an option to disable this behavior built in to only switch between 100% and FIT screen. If we need more than 100% zoom we can always use the ctrl key and the mouse wheel to do that.

I personally never want to zoom to more than 100% in darktable.

**Describe the solution you'd like**
I have this "dirty" hack in my personal build but you can of course implement it any way you see fit.

***FILE: darktable/src/views/darkroom.c:
FUNCTION: button_pressed(...)
LINE START: 3614***
```C
const gboolean cycle_zoom_200 = dt_conf_get_bool("darkroom/mouse/middle_button_cycle_zoom_to_200_percent");

if((tscale > 0.95f) && (tscale < 1.05f) && cycle_zoom_200) // we are at 100% and switch to 200%
{
  zoom = DT_ZOOM_1;
  scale = dt_dev_get_zoom_scale(dev, DT_ZOOM_1, 1.0, 0);
  if(low_ppd) closeup = 1;
}
else if(((tscale > 1.95f) && (tscale < 2.05f) && cycle_zoom_200) // Reset zoom switch to zoomfit
        || (zoom != DT_ZOOM_FIT && !cycle_zoom_200))
{
  zoom = DT_ZOOM_FIT;
  scale = dt_dev_get_zoom_scale(dev, DT_ZOOM_FIT, 1.0, 0);
}
else // other than 100 or 200% so zoom to 100 %
{
  if(low_ppd)
  {
      zoom = DT_ZOOM_1;
      scale = dt_dev_get_zoom_scale(dev, DT_ZOOM_1, 1.0, 0);
  }
  else
  {
    zoom = DT_ZOOM_FREE;
    scale = 1.0f / ppd;
  }
}
```
Happy coding from Denmark.

### PR DESCRIPTION
**Pull request by jmoric**
Fix #14046 

1. Added option for the user to change middle mouse button zoom level.
2. Removed redundant variables and calculations.
3. Some reformatting.

This is my first 'Pull request' so if I have made any mistake(s) let me know.

### CODE REVIEW

File: src/views/darkroom.c [hunk 1]
```diff
@@ -3707,31 +3708,37 @@ int button_pressed(dt_view_t *self,
...
     const gboolean low_ppd = (darktable.gui->ppd == 1);
-    const float mouse_off_x = x - 0.5f * dev->width;
-    const float mouse_off_y = y - 0.5f * dev->height;
-    zoom_x += mouse_off_x / (procw * scale);
-    zoom_y += mouse_off_y / (proch * scale);
     const float tscale = scale * ppd;
     closeup = 0;
-    if((tscale > 0.95f) && (tscale < 1.05f)) // we are at 100% and switch to 200%
+
+    const char *const key_zoom_level = "darkroom/mouse/middle_button_cycle_zoom_to_200_percent";
```
Comment (TurboGit) [inline src/views/darkroom.c:3725]: You should not need to check for if the key exists. The key should be added into data/darktableconfig.xml.in and assigned a default TRUE or FALSE. Probably TRUE to keep current behavior by default.

This configuration should be made visible to the user in the preference dialog.

So probably:

`<dtconfig prefs="darkroom" section="general">`

File: src/views/darkroom.c [hunk 2]
```diff
@@ -3707,31 +3708,37 @@ int button_pressed(dt_view_t *self,
...
+    // Get config so we can check if the user want to cycle through 100%->200%->FIT or
+    // only switch between FIT<->100%.
+    if(!dt_conf_key_exists(key_zoom_level))
+    {
+      dt_conf_set_bool(key_zoom_level, 1); // Set default TRUE
```
Comment (TurboGit) [inline src/views/darkroom.c:3730]: 1 -> TRUE

**Review by TurboGit — CHANGES_REQUESTED** [review]
Not tested yet.

File: src/views/darkroom.c (hunk 1 above)
Comment (jmoric) [reply src/views/darkroom.c]: I totally agree. I knew it was a bad place to run that kind of code on every middle click. I should have mentioned it in the description. Sorry!

To be honest, I looked through the code to find a better place to put that initialization code, but I totally missed out on the data folder. Yah, we learn every day.

I have corrected the code and it runs as intended.
Thanks for your time and guidance.

File: src/views/darkroom.c [hunk 3]
```diff
@@ -3707,31 +3708,37 @@ int button_pressed(dt_view_t *self,
...
-    zoom_x += mouse_off_x / (procw * scale);
-    zoom_y += mouse_off_y / (proch * scale);
     const float tscale = scale * ppd;
     closeup = 0;
-    if((tscale > 0.95f) && (tscale < 1.05f)) // we are at 100% and switch to 200%
+
+    const char *const key_zoom_level = "darkroom/mouse/middle_button_cycle_zoom_to_200_percent";
+    // Get config so we can check if the user want to cycle through 100%->200%->FIT or
+    // only switch between FIT<->100%.
+    if(!dt_conf_key_exists(key_zoom_level))
```
[the remark this answers is not in the document]
Comment (jmoric) [reply src/views/darkroom.c]: Done, see above.

**Comment by TurboGit** [comment]
@jmoric : I'll review this soon. Just a side comment we don't want to merge `master` into a **topic branch**. If needed a rebase on top of master is possible. Either you do that here and force push or I'll squash all the commits together anyway.

**Comment by jmoric** [comment]
@TurboGit : I understand. Sorry!
I could do a `git reset --hard HEAD~1` and push if you preferred that instead.

File: src/views/darkroom.c [hunk 4]
```diff
@@ -3707,31 +3708,32 @@ int button_pressed(dt_view_t *self,
...
     int closeup = dt_control_get_dev_closeup();
     float zoom_x = dt_control_get_dev_zoom_x();
     float zoom_y = dt_control_get_dev_zoom_y();
-    dt_dev_get_processed_size(dev, &procw, &proch);
     float scale = dt_dev_get_zoom_scale(dev, zoom, 1<<closeup, 0);
     const float ppd = darktable.gui->ppd;
     const gboolean low_ppd = (darktable.gui->ppd == 1);
-    const float mouse_off_x = x - 0.5f * dev->width;
-    const float mouse_off_y = y - 0.5f * dev->height;
-    zoom_x += mouse_off_x / (procw * scale);
```
Comment (TurboGit) [inline src/views/darkroom.c:3724]: I think removing this was not expected. At least it breaks the zoom center area. The goal was when zooming to have the same part of the image under the mouse cursor.

**Review by TurboGit — CHANGES_REQUESTED** [review]
Issue with the zoom center area.

File: src/views/darkroom.c (hunk 4 above)
Comment (jmoric) [reply src/views/darkroom.c:3724]: This is so embarrassing. I didn't catch that when I tested. I know I've screwed up. I am sorry!

I know I can `rebase -i` and delete the commit but I don't know how that's going to work out when the code is already uploaded.

I am sorry I'm taking your time on this, but I don't know how to handle this scenario in git/github with live code. Any suggestions?

Comment (TurboGit) [reply src/views/darkroom.c:3724]: Don't be sorry ! The good point is that you're getting better at fixing dt :)

For now, just fix the code and push the new commits there. TIA.

Comment (jmoric) [reply src/views/darkroom.c]: I 'm still sorry. I don't want to be kicked out. I have learned a lot already and I'll get there :)

I have reverted the commit and I hope it fulfills your request.

**Comment by TurboGit** [comment]
Working fine and this feels better for my use when I do not select zoom to 200.

Now for the rare case where I want to zoom more what about enabling the 200% zoom when the `ctrl` is selected while the middle mouse button is clicked. That is, if `ctrl` is pressed the `cycle_zoom_200` variable is TRUE.

The rational.

1. First you click middle button to check some details
2. You need a bit more zoom
3. Ctrl+Click + middle button and you go to 200%
4. Otherwise, just a middle click and we are back to fit

How does that sound?

**Comment by TurboGit** [comment]
@jmoric : See discussion in #14046, maybe we can even remove the preference. Let's see what others think about this.

**Comment by jmoric** [comment]
@TurboGit : Yes, that's sounds good. I suggest we keep the discussion to the original #14046 post and the implementation here so we don't get lost in cross-posting :)

**Comment by wpferguson** [comment]
Tested, works for me :smile:

**Comment by TurboGit** [comment]
Works for me, can you please provide a description for the release-notes? TIA.

**Review by TurboGit — APPROVED** [review]
Thanks!

**Comment by jmoric** [comment]
@TurboGit : Release-note:
Added the option to change the zoom behavior for the middle mouse button i darkroom. The setting can be enabled in the preference dialog under 'darkroom' category.

**Comment by TurboGit** [comment]
Thanks, I have put to be a bit more information:

> - Added the option to change the zoom behavior for the middle mouse
>   button in darkroom. The setting can be enabled in the preference
>   dialog under 'darkroom' category. The new behavior will only zoom
>   between fit and 100%. To have 200% zoom a <kbd>ctrl+click</kbd> is
>   needed.

Let me know if this is ok.

**Comment by jmoric** [comment]
Ha ha, yes, that's definitely a better release note. Thanks for correction.

**Pull request merged**